rust/cubestore/cubestore/src/cachestore/lazy.rs

```diff
@@ -0,0 +1,168 @@
+use crate::cachestore::{CacheItem, CacheStore, RocksCacheStore};
+use crate::config::ConfigObj;
+use crate::metastore::{IdRow, MetaStoreFs};
+use crate::CubeError;
+use async_trait::async_trait;
+use log::trace;
+use std::path::Path;
+use std::sync::Arc;
+use tokio::sync::watch::{Receiver, Sender};
+
+pub enum LazyRocksCacheStoreState {
+    FromRemote {
+        path: String,
+        metastore_fs: Arc<dyn MetaStoreFs>,
+        config: Arc<dyn ConfigObj>,
+        init_flag: Sender<bool>,
+    },
+    Initialized {
+        store: Arc<RocksCacheStore>,
+    },
+}
+
+pub struct LazyRocksCacheStore {
+    init_signal: Option<Receiver<bool>>,
+    state: tokio::sync::RwLock<LazyRocksCacheStoreState>,
+}
+
+impl LazyRocksCacheStore {
+    pub async fn load_from_dump(
+        path: &Path,
+        dump_path: &Path,
+        metastore_fs: Arc<dyn MetaStoreFs>,
+        config: Arc<dyn ConfigObj>,
+    ) -> Result<Arc<Self>, CubeError> {
+        let store = RocksCacheStore::load_from_dump(path, dump_path, metastore_fs, config).await?;
+
+        Ok(Arc::new(Self {
+            init_signal: None,
+            state: tokio::sync::RwLock::new(LazyRocksCacheStoreState::Initialized { store }),
+        }))
+    }
+
+    pub async fn load_from_remote(
+        path: &str,
+        metastore_fs: Arc<dyn MetaStoreFs>,
+        config: Arc<dyn ConfigObj>,
+    ) -> Result<Arc<Self>, CubeError> {
+        let (init_flag, init_signal) = tokio::sync::watch::channel::<bool>(false);
+
+        Ok(Arc::new(Self {
+            init_signal: Some(init_signal),
+            state: tokio::sync::RwLock::new(LazyRocksCacheStoreState::FromRemote {
+                path: path.to_string(),
+                metastore_fs,
+                config,
+                init_flag,
+            }),
+        }))
+    }
+
+    async fn init(&self) -> Result<Arc<RocksCacheStore>, CubeError> {
+        {
+            let guard = self.state.read().await;
+            if let LazyRocksCacheStoreState::Initialized { store } = &*guard {
+                return Ok(store.clone());
+            };
+        }
+
+        let mut guard = self.state.write().await;
+        match &*guard {
+            LazyRocksCacheStoreState::FromRemote {
+                path,
+                metastore_fs,
+                config,
+                // receiver will be closed on drop
+                init_flag: _,
+            } => {
+                let store =
```

Comment:
I tried to use move logic with initializing state & std::mem::replace, but I realized that load_from_remote may fail.